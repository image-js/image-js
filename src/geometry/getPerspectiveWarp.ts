import { Matrix, SingularValueDecomposition } from 'ml-matrix';

import type { Point } from '../utils/geometry/points.js';

/**
 * Options for getPerspectiveWarp function.
 */
export interface GetPerspectiveWarpOptions {
  /**
   * The horizontal dimension (in pixels) of the final rectified rectangular image.
   */
  width?: number;
  /**
   * The vertical dimension (in pixels) of the final rectified rectangular image.
   */
  height?: number;
}
/**
 * Returns result matrix along with vertical and horizontal dimensions for the rectangular image.
 */
export type GetPerspectiveWarpData = Required<GetPerspectiveWarpOptions> & {
  matrix: number[][];
};

// REFERENCES :
// http://graphics.cs.cmu.edu/courses/15-463/2011_fall/Lectures/morphing.pdf
/**
 * Computes a perspective transformation matrix to rectify a quadrilateral region into a rectangle.
 *
 * This function takes four corner points of a quadrilateral (e.g., a document photographed at an angle) and calculates the transformation matrix needed to "unwarp" it into a rectangular image.
 * The output dimensions can be specified or calculated automatically based on the input geometry.
 * @param pts - 4 reference corners.
 * @param options - PerspectiveWarpOptions.
 * @returns - Matrix from 4 points.
 */
export function getPerspectiveWarp(
  pts: Point[],
  options: GetPerspectiveWarpOptions = {},
): GetPerspectiveWarpData {
  if (pts.length !== 4) {
    throw new Error(
      `The array pts must have four elements, which are the four corners. Currently, pts have ${pts.length} elements`,
    );
  }
  const { width, height } = options;
  const [tl, tr, br, bl] = order4Points(pts);

  let widthRect;
  let heightRect;
  if (height && width) {
    widthRect = width;
    heightRect = height;
  } else if (!height && !width) {
    widthRect = Math.ceil(
      Math.max(distance2Points(tl, tr), distance2Points(bl, br)),
    );
    heightRect = Math.ceil(
      Math.max(distance2Points(tl, bl), distance2Points(tr, br)),
    );
  } else {
    throw new Error(
      `Invalid dimensions: ${width ? '`height`' : '`width`'} is missing. ` +
        `Either provide both width and height, or omit both to auto-calculate dimensions.`,
    );
  }

  const [x1, y1] = [0, 0];
  const [x2, y2] = [widthRect - 1, 0];
  const [x3, y3] = [widthRect - 1, heightRect - 1];
  const [x4, y4] = [0, heightRect - 1];

  const S = new Matrix([
    [x1, y1, 1, 0, 0, 0, -x1 * tl.column, -y1 * tl.column],
    [x2, y2, 1, 0, 0, 0, -x2 * tr.column, -y2 * tr.column],
    [x3, y3, 1, 0, 0, 0, -x3 * br.column, -y3 * br.column],
    [x4, y4, 1, 0, 0, 0, -x4 * bl.column, -y4 * bl.column],
    [0, 0, 0, x1, y1, 1, -x1 * tl.row, -y1 * tl.row],
    [0, 0, 0, x2, y2, 1, -x2 * tr.row, -y2 * tr.row],
    [0, 0, 0, x3, y3, 1, -x3 * br.row, -y3 * br.row],
    [0, 0, 0, x4, y4, 1, -x4 * bl.row, -y4 * bl.row],
  ]);
  const D = Matrix.columnVector([
    tl.column,
    tr.column,
    br.column,
    bl.column,
    tl.row,
    tr.row,
    br.row,
    bl.row,
  ]);

  const svd = new SingularValueDecomposition(S);
  const T = svd.solve(D).to1DArray(); // solve S*T = D
  T.push(1);

  const M = [];
  for (let i = 0; i < 3; i++) {
    const row = [];
    for (let j = 0; j < 3; j++) {
      row.push(T[i * 3 + j]);
    }
    M.push(row);
  }
  return { matrix: M, width: widthRect, height: heightRect };
}

/**
 * Sorts 4 points in order =>[top-left,top-right,bottom-right,bottom-left]. Input points must be in clockwise or counter-clockwise order.
 * @param pts - Array of 4 points.
 * @returns Sorted array of 4 points.
 */
export function order4Points(pts: Point[]) {
  let tl: Point;
  let tr: Point;
  let br: Point;
  let bl: Point;

  let minX = pts[0].column;
  let indexMinX = 0;

  for (let i = 1; i < pts.length; i++) {
    if (pts[i].column < minX) {
      minX = pts[i].column;
      indexMinX = i;
    }
  }

  let minX2 = pts[(indexMinX + 1) % pts.length].column;
  let indexMinX2 = (indexMinX + 1) % pts.length;

  for (let i = 0; i < pts.length; i++) {
    if (pts[i].column < minX2 && i !== indexMinX) {
      minX2 = pts[i].column;
      indexMinX2 = i;
    }
  }
  if (pts[indexMinX2].row < pts[indexMinX].row) {
    tl = pts[indexMinX2];
    bl = pts[indexMinX];
    if (indexMinX !== (indexMinX2 + 1) % 4) {
      tr = pts[(indexMinX2 + 1) % 4];
      br = pts[(indexMinX2 + 2) % 4];
    } else {
      tr = pts[(indexMinX2 + 2) % 4];
      br = pts[(indexMinX2 + 3) % 4];
    }
  } else {
    bl = pts[indexMinX2];
    tl = pts[indexMinX];
    if (indexMinX2 !== (indexMinX + 1) % 4) {
      tr = pts[(indexMinX + 1) % 4];
      br = pts[(indexMinX + 2) % 4];
    } else {
      tr = pts[(indexMinX + 2) % 4];
      br = pts[(indexMinX + 3) % 4];
    }
  }
  return [tl, tr, br, bl];
}
/**
 *  Calculates distance between points.
 * @param p1 - Point1
 * @param p2 - Point2
 * @returns distance between points.
 */
function distance2Points(p1: Point, p2: Point) {
  return Math.hypot(p1.column - p2.column, p1.row - p2.row);
}
