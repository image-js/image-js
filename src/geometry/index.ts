export * from './resize.js';
export * from './rotate.js';
export * from './transform.js';
export * from './transformRotate.js';
export type { Point } from '../utils/geometry/points.js';
