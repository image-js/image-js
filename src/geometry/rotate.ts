import { Image } from '../Image.js';

export type RotateAngle = 90 | 180 | 270 | -90 | -180 | -270;

/**
 * Rotates an image in multiples of 90 degrees.
 * @param image - The image to rotate.
 * @param angle - The angle to rotate the image by. Positive values rotate clockwise, negative values rotate counter-clockwise.
 * @returns - The rotated image.
 */
export function rotate(image: Image, angle: RotateAngle): Image {
  const newWidth = angle % 180 === 0 ? image.width : image.height;
  const newHeight = angle % 180 === 0 ? image.height : image.width;
  const newImage = Image.createFrom(image, {
    width: newWidth,
    height: newHeight,
  });

  if (angle === 90 || angle === -270) {
    for (let column = 0; column < image.width; column++) {
      for (let row = 0; row < image.height; row++) {
        for (let channel = 0; channel < image.channels; channel++) {
          newImage.setValue(
            newImage.width - row - 1,
            column,
            channel,
            image.getValue(column, row, channel),
          );
        }
      }
    }
  } else if (angle === 180 || angle === -180) {
    for (let column = 0; column < image.width; column++) {
      for (let row = 0; row < image.height; row++) {
        for (let channel = 0; channel < image.channels; channel++) {
          newImage.setValue(
            newImage.width - column - 1,
            newImage.height - row - 1,
            channel,
            image.getValue(column, row, channel),
          );
        }
      }
    }
  } else if (angle === 270 || angle === -90) {
    for (let column = 0; column < image.width; column++) {
      for (let row = 0; row < image.height; row++) {
        for (let channel = 0; channel < image.channels; channel++) {
          newImage.setValue(
            row,
            newImage.height - column - 1,
            channel,
            image.getValue(column, row, channel),
          );
        }
      }
    }
  } else {
    throw new RangeError(`invalid angle: ${angle}`);
  }

  return newImage;
}
