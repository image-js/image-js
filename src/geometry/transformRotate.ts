import type { Image, ImageCoordinates } from '../Image.js';
import type { Point } from '../utils/geometry/points.js';

import type { TransformOptions } from './transform.js';
import { transform } from './transform.js';

export interface TransformRotateOptions extends TransformOptions {
  /**
   * Specify the rotation center point as a predefined string or a Point.
   * @default `'center'`.
   */
  center?: ImageCoordinates | Point;
  /**
   * Scaling factor for the rotated image.
   *  @default `1`
   */
  scale?: number;
}

/**
 * Rotate an image anti-clockwise of a given angle.
 * @param image - Original image.
 * @param angle - Angle in degrees.
 * @param options - Rotate options.
 * @returns A new rotated image.
 */
export function transformRotate(
  image: Image,
  angle: number,
  options: TransformRotateOptions = {},
): Image {
  const { center = 'center', scale = 1, ...otherOptions } = options;

  let centerCoordinates;
  if (typeof center === 'string') {
    centerCoordinates = image.getCoordinates(center);
  } else {
    centerCoordinates = center;
  }
  const transformMatrix = getRotationMatrix(angle, centerCoordinates, scale);

  return transform(image, transformMatrix, otherOptions);
}

/**
 * Generates a rotation matrix for the given angle.
 * @param angle - Angle in degrees.
 * @param center - Center point of the image.
 * @param scale - Scaling factor.
 * @returns 3 x 3 rotation matrix.
 */
function getRotationMatrix(
  angle: number,
  center: Point,
  scale: number,
): number[][] {
  const angleRadians = (angle * Math.PI) / 180;
  const cos = scale * Math.cos(angleRadians);
  const sin = scale * Math.sin(angleRadians);
  return [
    [cos, sin, (1 - cos) * center.column - sin * center.row],
    [-sin, cos, sin * center.column + (1 - cos) * center.row],
    [0, 0, 1],
  ];
}
