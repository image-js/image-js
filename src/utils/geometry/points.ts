/**
 * Coordinates of a point in an image with the top-left corner being the reference point. The point is a [column, row] array.
 */
export type ArrayPoint = [column: number, row: number];

/**
 * Coordinates of a point in an image with the top-left corner being the reference point.
 */
export interface Point {
  /**
   * Point row.
   *
   */
  row: number;
  /**
   * Point column.
   *
   */
  column: number;
}

/**
 * Calculates a new point that is the difference p1 - p2.
 * @param p1 - First point.
 * @param p2 - Second Point.
 * @returns Difference between the two points.
 */
export function difference(p1: Point, p2: Point): Point {
  return { column: p1.column - p2.column, row: p1.row - p2.row };
}

/**
 * Calculates a new point that is the sum p1 + p2.
 * @param p1 - First point.
 * @param p2 - Second Point.
 * @returns Sum of the two points.
 */
export function sum(p1: Point, p2: Point): Point {
  return { column: p1.column + p2.column, row: p1.row + p2.row };
}

/**
 * Normalize a point (more precisely the vector from the origin to the point).
 * @param point - Point to normalize.
 * @returns - Normalized point.
 */
export function normalize(point: Point): Point {
  const length = Math.hypot(point.column, point.row);
  return { column: point.column / length, row: point.row / length };
}

/**
 * Rotate an array of points by an angle in radians.
 * The rotation is clockwise and the reference is (0,0).
 * @param radians - Angle in radians.
 * @param points - Source points.
 * @returns The points after rotation.
 */
export function rotate(radians: number, points: readonly Point[]): Point[] {
  const results: Point[] = [];
  const cos = Math.cos(radians);
  const sin = Math.sin(radians);
  for (const point of points) {
    results.push({
      column: cos * point.column - sin * point.row,
      row: sin * point.column + cos * point.row,
    });
  }
  return results;
}

/**
 * Dot product of 2 points assuming vectors starting from (0,0).
 * @param p1 - First point.
 * @param p2 - Second point.
 * @returns Dot product between the two vectors.
 */
export function dot(p1: Point, p2: Point) {
  return p1.column * p2.column + p1.row * p2.row;
}

/**
 * Round the coordinates of the point.
 * @param point - The point.
 * @returns Rounded coordinates of the point.
 */
export function round(point: Point): Point {
  return { column: Math.round(point.column), row: Math.round(point.row) };
}

/**
 * Sort an array of points by column then row.
 * @param points - Array of points to sort.
 * @returns Sorted points.
 */
export function sortByColumnRow(points: Point[]): Point[] {
  return points.slice().sort((point1, point2) => {
    if (point1.column < point2.column) return -1;
    if (point1.column > point2.column) return 1;
    return point1.row - point2.row;
  });
}
