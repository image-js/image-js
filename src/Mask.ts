import { BitDepth, Image } from './Image';
import { subtract, SubtractImageOptions } from './compare';
import {
  drawLineOnMask,
  DrawLineOnMaskOptions,
  drawPolygonOnMask,
  DrawPolygonOnMaskOptions,
  drawPolylineOnMask,
  DrawPolylineOnMaskOptions,
  drawRectangle,
  DrawRectangleOptions,
} from './draw';
import { drawPoints, DrawPointsOptions } from './draw/drawPoints';
import {
  and,
  AndOptions,
  invert,
  InvertOptions,
  or,
  OrOptions,
} from './filters';
import {
  GetBorderPointsOptions,
  getBorderPoints,
  getConvexHull,
  ConvexHull,
  Feret,
  getFeret,
} from './maskAnalysis';
import { getMbr, Mbr } from './maskAnalysis/getMbr';
import {
  bottomHat,
  BottomHatOptions,
  clearBorder,
  ClearBorderOptions,
  close,
  CloseOptions,
  dilate,
  DilateOptions,
  ErodeOptions,
  floodFill,
  FloodFillOptions,
  morphologicalGradient,
  MorphologicalGradientOptions,
  open,
  OpenOptions,
  solidFill,
  SolidFillOptions,
  topHat,
  TopHatOptions,
} from './morphology';
import { erode } from './morphology/erode';
import {
  convertColor,
  copyTo,
  CopyToOptions,
  paintMaskOnMask,
  PaintMaskOnMaskOptions,
} from './operations';
import { boolToNumber } from './utils/boolToNumber';
import { ImageColorModel, colorModels } from './utils/constants/colorModels';
import { Point } from './utils/geometry/points';

export type BitValue = 1 | 0 | boolean;

export interface MaskOptions {
  /**
   * Origin of the image relative to a parent image.
   * @default `{row: 0, column: 0 }`
   */
  origin?: Point;
  /**
   * Typed array holding the mask data.
   */
  data?: Uint8Array;
}

export interface MaskCreateFromOptions extends MaskOptions {
  width?: number;
  height?: number;
}

export class Mask {
  /**
   * The number of columns of the mask.
   */
  public readonly width: number;

  /**
   * The number of rows of the mask.
   */
  public readonly height: number;

  /**
   * The total number of bits in the mask (width Ã— height).
   */
  public readonly size: number;

  /**
   * The number of bits per value in each channel (always 1).
   */
  public readonly bitDepth: BitDepth;

  /**
   * The color model of the mask (always BINARY).
   */
  public readonly colorModel: ImageColorModel;

  /**
   * The number of color channels in the image, excluding the alpha channel (always 1).
   */
  public readonly components: number;

  /**
   * The number of channels in the mask, including the alpha channel (always 1).
   */
  public readonly channels: number;

  /**
   * Specifying that the mask has no an alpha channel.
   */
  public readonly alpha: boolean;

  /**
   * The maximum value that a pixel channel can have.
   */
  public readonly maxValue: number;

  /**
   * Origin of the image relative to a the parent image.
   */
  public readonly origin: Point;

  /**
   * Typed array holding the mask data.
   */
  private readonly data: Uint8Array;

  /**
   * Construct a new Mask knowing its dimensions.
   * @param width - Image width.
   * @param height - Image height.
   * @param options - Image options.
   */
  public constructor(width: number, height: number, options: MaskOptions = {}) {
    const { data, origin = { row: 0, column: 0 } } = options;

    if (width < 1 || !Number.isInteger(width)) {
      throw new RangeError(
        `width must be an integer and at least 1. Received ${width}`,
      );
    }

    if (height < 1 || !Number.isInteger(height)) {
      throw new RangeError(
        `height must be an integer and at least 1. Received ${height}`,
      );
    }

    this.width = width;
    this.height = height;
    this.size = width * height;
    this.bitDepth = 1;
    this.colorModel = 'BINARY';
    this.origin = origin;

    const colorModelDef = colorModels[this.colorModel];
    this.components = colorModelDef.components;
    this.alpha = colorModelDef.alpha;
    this.channels = colorModelDef.channels;

    this.maxValue = 1;

    if (data === undefined) {
      this.data = new Uint8Array(this.size);
    } else {
      const expectedLength = this.size * this.channels;
      if (data.length !== expectedLength) {
        throw new RangeError(
          `incorrect data size: ${data.length}. Expected ${expectedLength}`,
        );
      }
      this.data = data;
    }
  }

  /**
   * Create a new Mask based on the properties of an existing one.
   * @param other - Reference Mask.
   * @param options - Mask options.
   * @returns New mask.
   */
  public static createFrom(
    other: Mask | Image,
    options: MaskCreateFromOptions = {},
  ): Mask {
    const {
      width = other.width,
      height = other.height,
      origin = other.origin,
    } = options;
    return new Mask(width, height, { origin, ...options });
  }

  /**
   * Get a pixel of the mask.
   * @param column - Column index.
   * @param row - Row index.
   * @returns The pixel.
   */
  public getPixel(column: number, row: number): number[] {
    const result = [];
    const index = row * this.width + column;
    result.push(this.data[index]);
    return result;
  }

  /**
   * Set a pixel.
   * @param column - Column index.
   * @param row - Row index.
   * @param value - The pixel value.
   */
  public setPixel(column: number, row: number, value: number[]): void {
    const index = row * this.width + column;
    this.data[index] = value[0];
  }

  /**
   * Set a pixel to a given value if the coordinates are inside the mask.
   * @param column - Column index.
   * @param row - Row index.
   * @param value - New color of the pixel to set.
   */
  public setVisiblePixel(column: number, row: number, value: number[]): void {
    if (column >= 0 && column < this.width && row >= 0 && row < this.height) {
      this.setPixel(column, row, value);
    }
  }

  /**
   * Get a pixel using its index.
   * @param index - Index of the pixel.
   * @returns The pixel.
   */
  public getPixelByIndex(index: number): number[] {
    return [this.data[index]];
  }

  /**
   * Set a pixel using its index.
   * @param index - Index of the pixel.
   * @param value - New value of the pixel to set.
   */
  public setPixelByIndex(index: number, value: number[]): void {
    this.data[index] = value[0];
  }

  /**
   * Create a mask from an array of points.
   * @param width - Width of the mask.
   * @param height - Height of the mask.
   * @param points - Reference Mask.
   * @returns New mask.
   */
  public static fromPoints(
    width: number,
    height: number,
    points: Point[],
  ): Mask {
    const mask = new Mask(width, height);

    for (const point of points) {
      mask.setBit(point.column, point.row, 1);
    }

    return mask;
  }

  /**
   * Create a copy of this mask.
   * @returns The mask clone.
   */
  public clone(): Mask {
    return Mask.createFrom(this, { data: this.data.slice() });
  }

  /**
   * Get the value of a bit.
   * @param column - Column index.
   * @param row - Row index.
   * @returns The bit value.
   */
  public getBit(column: number, row: number): number {
    const index = row * this.width + column;
    return this.data[index];
  }

  /**
   * Set the value of a bit.
   * @param column - Column index.
   * @param row - Row index.
   * @param value - New bit value.
   */
  public setBit(column: number, row: number, value: BitValue): void {
    const index = row * this.width + column;
    // @ts-expect-error: we know that value is a boolean
    this.data[index] = value;
  }

  /**
   * Get the value of a bit using index.
   * @param index - Index of the pixel.
   * @returns Value of the bit.
   */
  public getBitByIndex(index: number): number {
    return this.data[index * this.channels];
  }

  /**
   * Set the value of a bit using index.
   * @param index - Index of the pixel.
   * @param value - Value to set.
   */
  public setBitByIndex(index: number, value: BitValue): void {
    const result = boolToNumber(value);
    this.data[index * this.channels] = result;
  }

  /**
   * Get the value of a bit. Function exists for compatibility with Image.
   * @param column - Column index.
   * @param row - Row index.
   * @param channel - Index of the channel, must be zero.
   * @returns The bit value.
   */
  public getValue(column: number, row: number, channel: number): number {
    checkChannel(channel);
    return this.getBit(column, row);
  }

  /**
   * Set the value of a bit. Function exists for compatibility with Image.
   * @param column - Column index.
   * @param row - Row index.
   * @param channel - Index of the channel, must be zero.
   * @param value - New bit value.
   */
  public setValue(
    column: number,
    row: number,
    channel: number,
    value: BitValue,
  ): void {
    checkChannel(channel);
    this.setBit(column, row, value);
  }

  /**
   * Get the value of a bit using index. Function exists for compatibility with Image.
   * @param index - Index of the pixel.
   * @param channel - Index of the channel, must be zero.
   * @returns Value of the bit.
   */
  public getValueByIndex(index: number, channel: number): number {
    checkChannel(channel);
    return this.getBitByIndex(index);
  }

  /**
   * Set the value of a bit using index. Function exists for compatibility with Image.
   * @param index - Index of the pixel.
   * @param channel - Index of the channel, must be zero.
   * @param value - Value to set.
   */
  public setValueByIndex(
    index: number,
    channel: number,
    value: BitValue,
  ): void {
    checkChannel(channel);
    this.setBitByIndex(index, value);
  }

  /**
   * Get the value of a specific bit. Select bit using a point.
   * @param point - Coordinates of the desired biz.
   * @returns Value of the bit.
   */
  public getValueByPoint(point: Point): number {
    return this.getValue(point.column, point.row, 0);
  }

  /**
   * Set the value of a specific bit. Select bit using a point.
   * @param point - Coordinates of the bit.
   * @param value - Value to set.
   */
  public setValueByPoint(point: Point, value: BitValue): void {
    this.setValue(point.column, point.row, 0, value);
  }

  /**
   * Return the raw mask data.
   * @returns The raw data.
   */
  public getRawImage() {
    return {
      width: this.width,
      height: this.height,
      data: this.data,
    };
  }

  public [Symbol.for('nodejs.util.inspect.custom')](): string {
    let dataString;
    if (this.height > 20 || this.width > 20) {
      dataString = '[...]';
    } else {
      dataString = printData(this);
    }
    return `Mask {
  width: ${this.width}
  height: ${this.height}
  data: ${dataString}
}`;
  }

  /**
   * Fill the mask with a value.
   * @param value - Value of the bit.
   * @returns The mask instance.
   */
  public fill(value: BitValue): this {
    const result = boolToNumber(value);
    this.data.fill(result);
    return this;
  }

  public convertColor(colorModel: ImageColorModel): Image {
    return convertColor(this, colorModel);
  }

  // FILTERS
  /**
   * Invert the colors of the mask.
   * @param options - Inversion options.
   * @returns The inverted mask.
   */
  public invert(options?: InvertOptions): Mask {
    return invert(this, options);
  }

  /**
   * Subtract other from a mask.
   * @param other - Image to subtract.
   * @param options - Inversion options.
   * @returns The subtracted mask.
   */
  public subtract(other: Mask, options?: SubtractImageOptions): Mask {
    return subtract(this, other, options);
  }

  /**
   * Perform an AND operation on two masks.
   * @param other - Second mask.
   * @param options - And options.
   * @returns AND of the two masks.
   */
  public and(other: Mask, options?: AndOptions): Mask {
    return and(this, other, options);
  }
  /**
   * Perform an OR operation on two masks.
   * @param other - Second mask.
   * @param options - And options.
   * @returns OR of the two masks.
   */
  public or(other: Mask, options?: OrOptions): Mask {
    return or(this, other, options);
  }

  // MASK ANALYSIS
  /**
   * Get the coordinates of the points on the border of a shape defined in a mask.
   * @param options - Get border points options.
   * @returns Array of border points.
   */
  public getBorderPoints(options?: GetBorderPointsOptions): Point[] {
    return getBorderPoints(this, options);
  }

  /**
   * Get the vertices of the convex Hull polygon of a mask.
   * @returns Array of the vertices of the convex Hull in clockwise order.
   */
  public getConvexHull(): ConvexHull {
    return getConvexHull(this);
  }

  /**
   * Get the corners of the minimum bounding rectangle of a shape defined in a mask.
   * @returns Array of border points.
   */
  public getMbr(): Mbr {
    return getMbr(this);
  }

  /**
   * Computes the Feret data.
   * @returns The Feret diameters.
   */
  public getFeret(): Feret {
    return getFeret(this);
  }

  // MORPHOLOGY
  /**
   * Erode a Mask.
   * @param options - Erode options.
   * @returns The eroded mask.
   */
  public erode(options?: ErodeOptions): Mask {
    return erode(this, options);
  }

  /**
   * Dilate an image.
   * @param options - Dilate options.
   * @returns The dilated image.
   */
  public dilate(options?: DilateOptions): Mask {
    return dilate(this, options);
  }

  /**
   * Open an image.
   * @param options - Open options.
   * @returns The opened image.
   */
  public open(options?: OpenOptions): Mask {
    return open(this, options);
  }

  /**
   * Close an image.
   * @param options - Close options.
   * @returns The closed image.
   */
  public close(options?: CloseOptions): Mask {
    return close(this, options);
  }

  /**
   * Top hat of an image.
   * @param options - Top hat options.
   * @returns The top-hatted image.
   */
  public topHat(options?: TopHatOptions): Mask {
    return topHat(this, options);
  }

  /**
   * Bottom hat of an image.
   * @param options - Bottom hat options.
   * @returns The bottom-hatted image.
   */
  public bottomHat(options?: BottomHatOptions): Mask {
    return bottomHat(this, options);
  }

  /**
   * Apply morphological gradient to an image.
   * @param options - Morphological gradient options.
   * @returns The processed image.
   */
  public morphologicalGradient(options?: MorphologicalGradientOptions): Mask {
    return morphologicalGradient(this, options);
  }

  /**
   * Remove elements connected to the borders of an image.
   * @param options - Clear border options.
   * @returns The processed image.
   */
  public clearBorder(options?: ClearBorderOptions): Mask {
    return clearBorder(this, options);
  }
  /**
   * Apply flood fill algorithm from a given starting point.
   * @param options - Flood fill options.
   * @returns The filled mask.
   */
  public floodFill(options?: FloodFillOptions): Mask {
    return floodFill(this, options);
  }
  /**
   * Fill holes in regions of interest.
   * @param options - Flood fill options.
   * @returns The filled mask.
   */
  public solidFill(options?: SolidFillOptions): Mask {
    return solidFill(this, options);
  }

  // DRAW

  /**
   * Draw a set of points on a mask.
   * @param points - Array of points.
   * @param options - Draw points on Image options.
   * @returns New mask.
   */
  public drawPoints(points: Point[], options: DrawPointsOptions = {}): Mask {
    return drawPoints(this, points, options);
  }

  /**
   * Draw a line defined by two points onto a mask.
   * @param from - Line starting point.
   * @param to - Line ending point.
   * @param options - Draw Line options.
   * @returns The mask with the line drawing.
   */
  public drawLine(
    from: Point,
    to: Point,
    options: DrawLineOnMaskOptions = {},
  ): Mask {
    return drawLineOnMask(this, from, to, options);
  }

  /**
   * Draw a polyline defined by an array of points on a mask.
   * @param points - Polyline array of points.
   * @param options - Draw polyline options.
   * @returns The mask with the polyline drawing.
   */
  public drawPolyline(
    points: Point[],
    options: DrawPolylineOnMaskOptions = {},
  ): Mask {
    return drawPolylineOnMask(this, points, options);
  }

  /**
   * Draw a polygon defined by an array of points onto an mask.
   * @param points - Polygon vertices.
   * @param options - Draw Line options.
   * @returns The mask with the polygon drawing.
   */
  public drawPolygon(
    points: Point[],
    options: DrawPolygonOnMaskOptions = {},
  ): Mask {
    return drawPolygonOnMask(this, points, options);
  }

  /**
   * Draw a rectangle defined by position of the top-left corner, width and height.
   * @param options - Draw rectangle options.
   * @returns The image with the rectangle drawing.
   */
  public drawRectangle(options: DrawRectangleOptions<Mask> = {}): Mask {
    return drawRectangle(this, options);
  }

  // OPERATIONS

  /**
   * Copy the mask to another one by specifying the location in the target mask.
   * @param target - The target mask.
   * @param options - Options.
   * @returns The target with the source copied to it.
   */
  public copyTo(target: Mask, options: CopyToOptions<Mask> = {}): Mask {
    return copyTo(this, target, options);
  }

  /**
   * Paint a mask onto another mask and the given position and with the given value.
   * @param mask - Mask to paint.
   * @param options - Paint mask options.
   * @returns The painted mask.
   */
  public paintMask(mask: Mask, options?: PaintMaskOnMaskOptions): Mask {
    return paintMaskOnMask(this, mask, options);
  }
}

/**
 * Returns all values of a mask as a string.
 * @param mask - Input mask.
 * @returns Formatted string with all values of a mask.
 */
function printData(mask: Mask): string {
  const result = [];
  for (let row = 0; row < mask.height; row++) {
    const line = [];
    for (let column = 0; column < mask.width; column++) {
      line.push(String(mask.getBit(column, row)));
    }
    result.push(`[${line.join(' ')}]`);
  }
  return result.join('\n        ');
}

/**
 * Verify the channel value of a mask.
 * @param channel - The channel value.
 */
function checkChannel(channel: number) {
  if (channel !== 0) {
    throw new RangeError(
      `channel value must be 0 on type Mask. Received ${channel}`,
    );
  }
}
