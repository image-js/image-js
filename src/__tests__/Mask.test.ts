import util from 'node:util';

import { Mask } from '../Mask';
import { Point } from '../utils/geometry/points';

describe('create new masks', () => {
  it('should create a mask', () => {
    const mask = new Mask(10, 20);
    expect(mask).toMatchObject({
      width: 10,
      height: 20,
      size: 200,
      bitDepth: 1,
      colorModel: 'BINARY',
      components: 1,
      channels: 1,
      alpha: false,
      maxValue: 1,
    });
    expect(mask.getRawImage().data).toHaveLength(200);
  });
  it('should throw on wrong width', () => {
    expect(() => new Mask(0, 1)).toThrow(
      /width must be an integer and at least 1. Received 0/,
    );
    expect(() => new Mask(0.5, 1)).toThrow(
      /width must be an integer and at least 1. Received 0.5/,
    );
  });
  it('should throw on wrong height', () => {
    expect(() => new Mask(1, 0)).toThrow(
      /height must be an integer and at least 1. Received 0/,
    );
    expect(() => new Mask(1, 0.5)).toThrow(
      /height must be an integer and at least 1. Received 0.5/,
    );
  });
  it('should throw on wrong data size', () => {
    const data = new Uint8Array(2);
    expect(() => new Mask(2, 2, { data })).toThrow(
      /incorrect data size: 2. Expected 4/,
    );
  });
});

describe('get and set bit', () => {
  it('should get and set', () => {
    const mask = new Mask(10, 20);
    expect(mask.getBit(5, 15)).toBe(0);
    mask.setBit(5, 15, 1);
    expect(mask.getBit(5, 15)).toBe(1);
  });
  it('should get and set by index', () => {
    const mask = new Mask(10, 20);
    expect(mask.getBitByIndex(15)).toBe(0);
    mask.setBitByIndex(15, 1);
    expect(mask.getBitByIndex(15)).toBe(1);
  });
  it('should get and set value by point', () => {
    const point = { column: 15, row: 0 };
    const mask = new Mask(10, 20);
    expect(mask.getValueByPoint(point)).toBe(0);
    mask.setValueByPoint(point, 1);
    expect(mask.getValueByPoint(point)).toBe(1);
    expect(mask.getNbNonZeroPixels()).toBe(1);
  });
});

describe('get and set value', () => {
  it('should get and set', () => {
    const mask = new Mask(10, 20);
    expect(mask.getValue(5, 15, 0)).toBe(0);
    mask.setValue(5, 15, 0, 1);
    expect(mask.getValue(5, 15, 0)).toBe(1);
  });
  it('should get and set by index', () => {
    const mask = new Mask(10, 20);
    expect(mask.getValueByIndex(15, 0)).toBe(0);
    mask.setValueByIndex(15, 0, 1);
    expect(mask.getValueByIndex(15, 0)).toBe(1);
  });
  it('wrong channel value', () => {
    const mask = new Mask(10, 20);
    expect(() => {
      mask.getValue(2, 1, 2);
    }).toThrow(/channel value must be 0 on type Mask. Received 2/);
  });
});

describe('get and set pixels', () => {
  it('should get and set', () => {
    const img = new Mask(10, 20);
    expect(img.getPixel(5, 15)).toStrictEqual([0]);
    img.setPixel(5, 15, [1]);
    expect(img.getPixel(5, 15)).toStrictEqual([1]);
  });

  it('should get and set by index', () => {
    const img = new Mask(10, 20);
    expect(img.getPixelByIndex(5)).toStrictEqual([0]);
    img.setPixelByIndex(5, [1]);
    expect(img.getPixelByIndex(5)).toStrictEqual([1]);
  });
});

test('fill with a value', () => {
  const mask = new Mask(2, 2);
  mask.fill(1);
  expect(mask).toMatchMaskData([
    [1, 1],
    [1, 1],
  ]);
  expect(mask.getNbNonZeroPixels()).toBe(4);
});

test('createFrom', () => {
  const mask = new Mask(2, 20);

  const newMask = Mask.createFrom(mask);
  expect(mask.width).toBe(newMask.width);
  expect(mask.height).toBe(newMask.height);
  expect(mask.origin).toStrictEqual(newMask.origin);
});

test('fromPoints', () => {
  const points: Point[] = [{ column: 1, row: 1 }];
  expect(Mask.fromPoints(5, 3, points)).toMatchMaskData([
    [0, 0, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ]);
});

test('clone', () => {
  const mask = new Mask(2, 2);
  mask.setBit(1, 0, 1);
  const copy = mask.clone();
  expect(copy).toMatchMask(mask);
});

test('check custom inspect', () => {
  const mask = new Mask(1, 2);

  expect(util.inspect(mask)).toMatchSnapshot();
});

test('check custom inspect with image too large', () => {
  const image = new Mask(25, 25);
  expect(util.inspect(image)).toMatchSnapshot();
});
