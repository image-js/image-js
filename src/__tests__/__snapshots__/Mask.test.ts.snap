// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`check custom inspect 1`] = `
"Mask {
  width: 1
  height: 2
  data: [0]
        [0]
}"
`;

exports[`check custom inspect with image too large 1`] = `
"Mask {
  width: 25
  height: 25
  data: [...]
}"
`;
